### PR TITLE
Devided #3856, split Space Cadet process, trying fix at #3701

### LINKED ISSUE #3701: Space Cadet Shifts sends extraneous input with KC_GESC
When typing with normal alphanumeric characters a quick shift + character sends the correct output. However, when space cadet shifts are enabled, Using either shifts for a `Shift + GESC` to access ~ outputs `~)` and `~(` when the keycombo is pressed quickly enough.

For reference, I'm using a DZ60 with pretty vanilla firmware (no special macros or anything).

Let me know if there's any additional information needed, but I think this should be reproducible in all keyboards.

### LINKED ISSUE #3769: GRAVE_ESC with GUI sends GUI+GRAVE, wanting just GRAVE
I was setting up my firmware and thought that was kind of strange. I have a 60%, and the desired behaviour was:

Esc = Escape
Shift+Esc = Tilde
Gui+Esc = Grave

But it sends Gui+Grave, which doesn't print the character. Is there a way to remedy this?

### PR DESCRIPTION
**Pull request by leico**
update: 09/12/18

* based on #3856 pull requests. 
* split Space Cadet process from `quantum/quantum.c` to `quantum/process_keycode/process_space_cadet.c/h`
* refactored Space Cadet process
    * try to fix #3701 
* ~~try to fix #3769~~
    * ~~defined `GESC_GR_IGNORED_MOD`~~
    * ~~defined `GESC_GR_IGNORED_MOD_BIT`~~

Todo:

* [ ] ~~`NO_SPACE_CADET` or~~ `SPACE_CADET_ENABLE`
    * [x] implement
    * [x] adapt default keymaps
    * [ ] review
* [ ] change manuals
    * [x] ~~ESC_GRAVE~~
    * [x] SPACE_CADET_SHIFT
    * [x] SHIFT_ENTER
    * [ ] need review (I'm not English native.)
* [ ] delete old sourced(comment outs)
* [ ] add comments if needed

### CODE REVIEW

**Comment by leico** [comment]
OK, try it as soon as possible.

File: quantum/quantum.c [hunk 1]
```diff
@@ -714,6 +752,54 @@ bool process_record_quantum(keyrecord_t *record) {
...
+          if( get_mods() & MOD_BIT( GESC_GR_IGNORED_MOD ) ) {
+            unregister_mods( MOD_BIT( GESC_GR_IGNORED_MOD ) );
+            enabled = true;
+          }
+#       elif defined GESC_GR_IGNORED_MOD_BIT
+          uint8_t mods = get_mods() & GESC_GR_IGNORED_MOD_BIT;
+          if( mods ) {
+            unregister_mods( mods );
+          }
+#       endif
```
Comment (skullydazed) [inline quantum/quantum.c]: As we talked about on discord please remove these special cases.

File: quantum/quantum_keycodes.h [hunk 2]
```diff
@@ -425,6 +425,7 @@ enum quantum_keycodes {
     RGB_MODE_GRADIENT,
     RGB_MODE_RGBTEST,
 
+#ifdef SPACE_CADET_ENABLE
```
Comment (skullydazed) [inline quantum/quantum_keycodes.h]: Please remove these. We want the keycodes to be defined even if they don't do anything.

File: keyboards/alu84/keymaps/default/rules.mk [hunk 3]
```diff
@@ -0,0 +1 @@
+SPACE_CADET_ENABLE = yes
```
[the remark this answers is not in the document]
Comment (leico) [reply keyboards/alu84/keymaps/default/rules.mk]: My worst mystake in this time is I don’t ask.

So, first I had to ask which is better, enable normal or disable normal. 

If enable normal,

* normal binary is bigger than disable
* don’t need change if keymap has already used

If disable normal,

* less binary size, provide more user space
* need change all keymaps, they have already used
* I don’t know which versioning rules use QMK, but perhaps need increase version number

[the remark this answers is not in the document]
Comment (leico) [reply keyboards/alu84/keymaps/default/rules.mk]: Increase mistake.... fix inccorect description

**Comment by leico** [comment]
interim pull request...

* [x] revert grave escape
    * [x] doc
    * [x] quantum.c
* [x] improve docs
    * [x] space cadet shift
* [ ] issue of enable default or disable default
    * [ ] change keymaps setting (if needed)
        * [x] [keyboard]/default keymap
        * [ ] layouts directory

File: keyboards/alu84/keymaps/default/rules.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (drashna) [reply keyboards/alu84/keymaps/default/rules.mk]: It should be `_ENABLE`, as "disabled by default" is the recommended config:
https://docs.qmk.fm/#/contributing?id=quantumtmk-core

File: docs/feature_space_cadet_shift.md [hunk 4]
```diff
@@ -4,14 +4,20 @@ Steve Losh described the [Space Cadet Shift](http://stevelosh.com/blog/2012/10/a
 
 ## Usage
 
-Replace the Left Shift key in your keymap with `KC_LSPO` (Left Shift, Parenthesis Open), and Right Shift with `KC_RSPC` (Right Shift, Parenthesis Close).
+First, you need to enable Space Cadet Shift feature. Add a flag in your `(keymap|keyboard)/rules.mk` like this:
```
[the remark this answers is not in the document]
Comment (leico) [reply docs/feature_space_cadet_shift.md]: Thank you!

File: keyboards/alu84/keymaps/default/rules.mk (hunk 3 above)
[the remark this answers is not in the document]
Comment (leico) [reply keyboards/alu84/keymaps/default/rules.mk]: Ah, OK. Could I change all layouts `rules.mk`, already used space cadet shift for this pull requests?

[the remark this answers is not in the document]
Comment (drashna) [reply keyboards/alu84/keymaps/default/rules.mk]: I would say leave it off for all of them. 

But it should be added to the template, even if disabled.  
https://github.com/qmk/qmk_firmware/blob/master/quantum/template/avr/rules.mk

**Comment by leico** [comment]
@drashna , Thank you reply.

Perhaps all of space cadet used layouts would enable, also template goes disable default.
one more, compare binary size between enable and disable, noted in documents how many bytes increase when enable.

**Comment by drashna** [comment]
This PR and #5226 aim to do the same thing. Though, that PR includes some additional functionality.

**Comment by leico** [comment]
@drashna yes, It looks similar directionality enhancements, #5226 looks more multifunctional. I am still waiting a result about internal structure discussion. 4 month ago, I and @skullydazed afraid to QMK souces become a `#define` , `#ifdef` spagetti hell.

**Comment by drashna** [comment]
I hate to say it, but I think that the space cadet feature has been overhauled, and also meets most of the same goals that this PR had, plus more.   

That's why there are a bunch of merge conflicts now. 

I'm closing this for now.  If you don't feel that's correct/appropriate, then please do re-open, and update your PR.

**Pull request closed without merging**